### PR TITLE
lds api: more error handling

### PR DESCRIPTION
**Pull request by mattklein123**
Check that the lds cluster exists and local info is available.

Fixes https://github.com/lyft/envoy/issues/1264

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team 

cc @stolsma

File: source/server/lds_api.cc [hunk 1]
```diff
@@ -16,6 +16,14 @@ LdsApi::LdsApi(const Json::Object& config, Upstream::ClusterManager& cm,
                      std::chrono::milliseconds(config.getInteger("refresh_delay_ms", 30000))),
       local_info_(local_info), listener_manager_(lm),
       stats_({ALL_LDS_STATS(POOL_COUNTER_PREFIX(scope, "listener_manager.lds."))}) {
+  if (local_info.clusterName().empty() || local_info.nodeName().empty()) {
+    throw EnvoyException("lds: setting --service-cluster and --service-node are required");
```
Comment (htuch) [inline source/server/lds_api.cc]: I notice we do this in other places, like CDS too. Why not refactor to push this check much earlier to some common path?

Comment (mattklein123) [inline source/server/lds_api.cc]: I was being lazy. Let me see if there is a good place to put it, or at least in a utility function.

**Comment by RomanDzhabarov** [comment]
LGTM after addressing htuch comments. +1

**Comment by mattklein123** [comment]
@htuch @RomanDzhabarov updated

File: source/common/router/rds_impl.cc [hunk 2]
```diff
@@ -56,14 +55,7 @@ RdsRouteConfigProviderImpl::RdsRouteConfigProviderImpl(
...
 
-  if (local_info.clusterName().empty() || local_info.nodeName().empty()) {
-    throw EnvoyException("rds: setting --service-cluster and --service-node are required");
-  }
-
-  if (!cm.get(remote_cluster_name_)) {
-    throw EnvoyException(fmt::format("rds: unknown remote cluster '{}'", remote_cluster_name_));
-  }
-
+  ::Envoy::Config::Utility::checkClusterAndLocalInfo("rds", remote_cluster_name_, cm, local_info);
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: I notice you use both fully qualified and partially qualified namespace prefixes, should there be some consistent style here?

File: source/common/config/utility.cc [hunk 3]
```diff
@@ -2,9 +2,30 @@
...
+                                       Upstream::ClusterManager& cm,
+                                       const LocalInfo::LocalInfo& local_info) {
+  if (!cm.get(cluster_name)) {
+    throw EnvoyException(fmt::format("{}: unknown cluster '{}'", error_prefix, cluster_name));
+  }
+
+  checkLocalInfo(error_prefix, local_info);
+}
+
+void Utility::checkLocalInfo(const std::string& error_prefix,
```
Comment (htuch) [inline source/common/config/utility.cc]: Out of curiosity, why can't this be done early on during initialization? Can local_info change dynamically?

File: source/common/router/rds_impl.cc (hunk 2 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: Prefr partially qualified, but in this case there is a namespace conflict so it's required.

File: source/common/config/utility.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/config/utility.cc]: The "problem" is that not all configs require local info to be populated. So --service-node and --server-cluster are not currently required. We could make them required, but that will break some people. This is probably the next best thing.

**Pull request merged**